### PR TITLE
Release ServiceFabric 2017-07-01-preview swagger specification

### PR DESCRIPTION
**Pull request by juhacket**
This change introduces applicationType, version, application, and service ARM resources for the ServiceFabric resource provider.

### CODE REVIEW

**Review by sergey-shandar — CHANGES_REQUESTED** [review]
@juhacket please, make changes in [readme.md](https://github.com/Azure/azure-rest-api-specs/blob/current/specification/servicefabric/resource-manager/readme.md)

File: specification/servicefabric/resource-manager/readme.md
Comment (sergey-shandar) [inline specification/servicefabric/resource-manager/readme.md]: Thank you for adding a readme file. Please, add a new tag and keep `package-2016-09` untouched. This way, we can generate SDK for different versions on demand and keep track of how we did it before.

**Review by sergey-shandar — CHANGES_REQUESTED** [review]

**Comment by sergey-shandar** [comment]
@ravbhatnagar it's a new API version for Service Fabric.

File: specification/servicefabric/resource-manager/readme.md
Comment (juhacket) [reply specification/servicefabric/resource-manager/readme.md]: Done

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (sergey-shandar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: use `true` instead of '"true"` for Booleans.

**Review by sergey-shandar — CHANGES_REQUESTED** [review]

File: specification/servicefabric/resource-manager/readme.md
Comment (sergey-shandar) [reply specification/servicefabric/resource-manager/readme.md]: @juhacket thank you, I will keep posting mini-review to fix CI until ARM approval. After that, I will send full review.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (juhacket) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: We had this document validated by the ARM team already. Does it need to be reviewed again?

Comment (juhacket) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: We are trying to publish it for Ignite.

Comment (sergey-shandar) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: @juhacket do you have an email/link to the validation? Another option (to make it faster) is to contact @ravbhatnagar directly so he label the PR as ARM signed off.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: ISO 8601 standard. Format should be marked as date-time.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: ISO 8601 standard. Format should be marked as date-time.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: ISO 8601 standard.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: This is existing API i think, but 201 + provisioningState + AzureAsyncOperation header is recommended for PUT  LRO instead of 202.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: It looks like all these state transitions happen during cluster provisioning. These values could be/should be merged with provisioningState. Existing API

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: How does customer discover the list of environments?

Comment (ravbhatnagar) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: environments is a property on the response of clusterVersions. This could be supported using a query parameter instead of modeling it as a resource. So to get cluster versions for a specific environment, customer will use this API -
/subscriptions/{subscriptionId}/providers/Microsoft.ServiceFabric/locations/{location}/clusterVersions and then do a $filter.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: What is updatable through this patch? There is nothing in the properties bag.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: Isnt applicationTypes a proxy resource? Proxy resources dont have a location, but in the response of this API, i see you have a location. Is it a tracked resource?

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: In application resource, you are refering the applicationTyperesource. It should be done using the arm resource id of the applicationTypeResource.

Comment (ravbhatnagar) [inline specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: versions resource is being referenced here. It should be done using the arm resource id.

**Review by ravbhatnagar — COMMENTED** [review]
@juhacket - Some comments for you. Some of it minor and some that may need to be revisited (specially the one where you are referencing type and version from Application resource). I know this would already have been implemented in service. Can we get an ACK from you that you can take a look at addressing the feedback in the next APi-version if it seems like the right design?

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (juhacket) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: We were planning to make it a tracked resource, but decided against it. We left this as is in case we changed our mind.

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (juhacket) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: Nothing since we changed this to a proxy resource.

Comment (ravbhatnagar) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: Should we remove this API then?

File: specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json
Comment (juhacket) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: Cluster doesn't support async operations yet. This will be in another API version.

Comment (ravbhatnagar) [reply specification/servicefabric/resource-manager/Microsoft.ServiceFabric/2017-07-01-preview/servicefabric.json]: Lets mark "x-ms-long-running-operation": "true" as false then. It is currently set to true. And we can also remove the 202 from the response.

**Comment by ravbhatnagar** [comment]
@sergey-shandar - lets merge this.  As you might have seen on the email, we are conditionally merging this as this is an Ignite deliverable but the team will address the feedback post Ignite. And also this is preview Api-version. 
@salameer -another conditionally merged PR.

**Comment by juhacket** [comment]
@sergey-shandar Can you commit this change? It looks like all requirements have been met.

**Comment by sergey-shandar** [comment]
@juhacket @ravbhatnagar we need to fix at least errors which prevents us to generate Node.js SDK. In particular, convert string to bool.

**Review by sergey-shandar — APPROVED** [review]
There are still some issues but we decided to merged this with the condition that the issues will be fixed ASAP.

File: specification/servicefabric/resource-manager/readme.md
Comment (lmazuel) [inline specification/servicefabric/resource-manager/readme.md]: @juhacket @sergey-shandar 
This is a breaking change and redefine the Swagger that correctly says `ServiceFabricManagementClient`

**Review by lmazuel — COMMENTED** [review]
Breaking change

File: specification/servicefabric/resource-manager/readme.md
Comment (sergey-shandar) [reply specification/servicefabric/resource-manager/readme.md]: @juhacket could you send a PR to fix this problem ASAP. @lmazuel I missed this line which effects all versions of API.

Comment (juhacket) [reply specification/servicefabric/resource-manager/readme.md]: https://github.com/Azure/azure-rest-api-specs/pull/1748

**Comment by amarzavery** [comment]
@sergey-shandar @juhacket - This PR has issues.
The global parameters should have the "x-ms-parameter-location": "method" extension if they do not intend to be properties on the client.

The constructor signature of the client changed **from** 
`constructor(credentials, subscriptionId, baseUri, options)`

**to**

`constructor(credentials, applicationName, applicationTypeName, serviceName, subscriptionId, version, cluster, clusterUpdateParameters, baseUri, options)`
which is a breaking and unnecessary change.
/cc @salameer 
https://github.com/Azure/azure-openapi-validator/issues/84

**Pull request merged**
